src/test/java/org/broadinstitute/hellbender/utils/pileup/ReadPileupUnitTest.java

```diff
@@ -0,0 +1,345 @@
+package org.broadinstitute.hellbender.utils.pileup;
+
+import htsjdk.samtools.CigarElement;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMReadGroupRecord;
+import htsjdk.samtools.util.Locatable;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.read.ArtificialReadUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static htsjdk.samtools.SAMRecord.NO_MAPPING_QUALITY;
+
+/**
+ * Test routines for read-backed pileup.
+ */
+public final class ReadPileupUnitTest {
+    static SAMFileHeader header;
+    private Locatable loc;
+
+    @BeforeClass
+    public void beforeClass() {
+        header = ArtificialReadUtils.createArtificialSamHeader(1, 1, 1000);
+        loc = new SimpleInterval("chr1", 1, 1);
+    }
+
+    /**
+     * Ensure that basic read group splitting works.
+     */
+    @Test
+    public void testSplitByReadGroup() {
+        final SAMReadGroupRecord readGroupOne = new SAMReadGroupRecord("rg1");
+        final SAMReadGroupRecord readGroupTwo = new SAMReadGroupRecord("rg2");
+
+        final SAMFileHeader header = ArtificialReadUtils.createArtificialSamHeader(1,1,1000);
+        header.addReadGroup(readGroupOne);
+        header.addReadGroup(readGroupTwo);
+
+        final GATKRead read1 = ArtificialReadUtils.createArtificialRead(header,"read1",0,1,10);
+        read1.setAttribute("RG",readGroupOne.getId());
```

Comment:
This is not the way to set read groups in the `GATKRead` interface (will only work for SAMRecord-backed GATKReads!). Use `setReadGroup()` instead (here and below).